From 9634ee5e28a91eb9258feeeb6cecf10cc38d15d2 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Thu, 23 Feb 2017 14:17:45 +0800
Subject: [PATCH 1537/5242] MLK-14252-1 video: mipi_dsi_northwest: reset dsi
 domains when shutdown.

commit  fcaf495c6484469e4de2fb6e84abdd0929f848fc from
https://source.codeaurora.org/external/imx/linux-imx.git

Put the three dsi domains into reset state before dsi is
shutdown to avoid dsi being a unstable state.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/mipi_dsi_northwest.c |   29 +++++++++++++++++---------
 1 file changed, 19 insertions(+), 10 deletions(-)

diff --git a/drivers/video/fbdev/mxc/mipi_dsi_northwest.c b/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
index 83129e8..6aa31bb 100644
--- a/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
+++ b/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
@@ -429,6 +429,20 @@ static int mipi_display_exit_sleep(struct mxc_dispdrv_handle *disp)
 	return err;
 }
 
+static void reset_dsi_domains(struct mipi_dsi_info *mipi_dsi, bool reset)
+{
+	/* escape domain */
+	regmap_update_bits(mipi_dsi->regmap, SIM_SOPT1CFG,
+			DSI_RST_ESC_N, (reset ? 0 : DSI_RST_ESC_N));
+	/* byte domain */
+	regmap_update_bits(mipi_dsi->regmap, SIM_SOPT1CFG,
+			DSI_RST_BYTE_N, (reset ? 0 : DSI_RST_BYTE_N));
+
+	/* dpi domain */
+	regmap_update_bits(mipi_dsi->regmap, SIM_SOPT1CFG,
+			DSI_RST_DPI_N, (reset ? 0 : DSI_RST_DPI_N));
+}
+
 static int mipi_dsi_enable(struct mxc_dispdrv_handle *disp,
 			   struct fb_info *fbi)
 {
@@ -461,16 +475,7 @@ static int mipi_dsi_enable(struct mxc_dispdrv_handle *disp,
 
 		mipi_dsi_dpi_init(mipi_dsi);
 
-		/* escape domain */
-		regmap_update_bits(mipi_dsi->regmap, SIM_SOPT1CFG,
-				DSI_RST_ESC_N, DSI_RST_ESC_N);
-		/* byte domain */
-		regmap_update_bits(mipi_dsi->regmap, SIM_SOPT1CFG,
-				DSI_RST_BYTE_N, DSI_RST_BYTE_N);
-
-		/* dpi domain */
-		regmap_update_bits(mipi_dsi->regmap, SIM_SOPT1CFG,
-				DSI_RST_DPI_N, DSI_RST_DPI_N);
+		reset_dsi_domains(mipi_dsi, 0);
 
 		/* display_en */
 		regmap_update_bits(mipi_dsi->regmap, SIM_SOPT1CFG,
@@ -882,6 +887,10 @@ static void mipi_dsi_shutdown(struct platform_device *pdev)
 
 		mipi_dsi->lcd_inited = 0;
 	}
+
+	reset_dsi_domains(mipi_dsi, 1);
+	regmap_update_bits(mipi_dsi->regmap, SIM_SOPT1CFG,
+			   DSI_PLL_EN, 0x0);
 }
 
 static const struct of_device_id imx_mipi_dsi_dt_ids[] = {
-- 
1.7.9.5

